src/persist/src/indexed/runtime.rs

```diff
@@ -0,0 +1,233 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! WIP
+
+#![allow(missing_docs)]
+
+use std::sync::mpsc::{Receiver, RecvError, SendError};
+use std::sync::{mpsc, Arc, Mutex};
+use std::thread::{self, JoinHandle};
+
+use log;
+
+use crate::error::Error;
+use crate::indexed::{Indexed, IndexedSnapshot};
+use crate::storage::{Blob, Buffer};
+use crate::Id;
+
+enum Cmd {
+    Register(Id, Box<dyn FnOnce(Result<(), Error>) + Send + 'static>),
```

Comment:
I'm not thrilled with how all these callbacks work atm. I could just pass channels around directly, but one thing that I like about the callbacks is that it's easy to keep the semantics around who's allowed to hang up in one place (example below). OTOH it's a lot of boxing and channels, which feels redundant. Nikhil, I'll take a look at what coord does, but also curious what your larger thoughts are.

```
        let (tx, rx) = mpsc::channel();
        let callback = move |res| {
            tx.send(res).expect("internal error");
        };
        let cmd = Cmd::Seal(id, ts, Box::new(callback));
        core.send(cmd);

        rx.recv().expect("this happens when the runtime shut down after the request queued but before it was processed")
```

Reply:
~Strong +1 on wrapping up channels in something else. One option is just to wrap them all in a newtype that enforces the correct hangup semantics but otherwise provides access to `send` and `recv`.~

~coord doesn't even expose the fact that it uses channels at all. You just call a normal-looking `async` function that happens to communicate with the coordinator under the hood: https://dev.materialize.com/api/rust/coord/struct.SessionClient.html~

~Typing out the `SessionClient` was a lot of boring nonsense but it makes the call sites soo much clearer.~

Ah, ok, so `Runtime` is like `coord::SessionClient`. The coordinator just stores the concrete `tx`. When it's all crate-internal it doesn't feel so bad to just casually have txn semantics. But! You could newtype the `tx` and `rx`—make a `PersistCmdTransmitter` or somesuch—and then that could enforce the hangup protocol you're enforcing. Is that... an answer to the question you were asking?